### PR TITLE
Feature: Add new gcode preview

### PR DESCRIPTION
**Pull request by bbaa77770**

### CODE REVIEW

File: src/app/flux/printing/index.js [hunk 1]
```diff
@@ -3103,7 +3251,9 @@ export const actions = {
 export default function reducer(state = INITIAL_STATE, action) {
     switch (action.type) {
         case ACTION_UPDATE_STATE: {
-            return Object.assign({}, state, action.state);
+            const s = Object.assign({}, state, action.state);
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:3254]: restore this

File: src/app/models/ModelGroup.js [hunk 2]
```diff
@@ -1874,6 +1874,12 @@ class ModelGroup extends EventEmitter {
         this.modelChanged();
     }
 
+    setDisplayType(displayedType) {
+        this.models.forEach(model => {
```
Comment (xiekun1992) [inline src/app/models/ModelGroup.js:1878]: including Prime Tower?

File: src/app/ui/widgets/PrintingVisualizer/ModeToggleBtn.jsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                            onToggleMode(value);
+                        }}
+                    >
+
+                        {
+                            GCODEPREVIEWMODES.map((item, index) => {
+                                return (
+                                    <div
+                                        key={item}
+                                        style={{ margin: '16px', marginRight: '10px' }}
```
Comment (xiekun1992) [inline src/app/ui/widgets/PrintingVisualizer/ModeToggleBtn.jsx:65]: define a CSS class?

File: src/app/ui/widgets/PrintingVisualizer/ModeToggleBtn.jsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                                        className="sm-flex justify-space-between height-24"
+                                    >
+                                        <div className="font-roboto font-weight-normal font-size-middle">
+                                            <SvgIcon
+                                                name={GCODEPREVIEWMODES_ICONS[index]}
+                                                size={24}
+                                                type={['static']}
+                                            />
+
+                                            <span style={{ fontSize: '14px' }} className="v-align-m margin-left-8">
```
Comment (xiekun1992) [inline src/app/ui/widgets/PrintingVisualizer/ModeToggleBtn.jsx:75]: use CSS class for `fontSize: '14px'`

File: src/app/workers/GcodeToBufferGeometry/gcodeBufferGeometryToObj3d.js [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
+// import { Line2 } from 'three/examples/jsm/lines/Line2';
 import { PRINT3D_UNIFORMS, PRINT3D_VERT_SHADER, PRINT3D_FRAG_SHADER } from '../ShaderMaterial/print3d-shader-meterial';
 import { WORKSPACE_UNIFORMS, WORKSPACE_FRAG_SHADER, WORKSPACE_VERT_SHADER } from '../ShaderMaterial/workspace-shader-meterial';
 
-
 const gcodeBufferGeometryToObj3d = (func, bufferGeometry, renderMethod) => {
     let obj3d = null;
     switch (func) {
         case '3DP':
+            console.log('gcodeBufferGeometryToObj3d');
```
Comment (xiekun1992) [inline src/app/workers/GcodeToBufferGeometry/gcodeBufferGeometryToObj3d.js:10]: remove log

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

**Comment by jane-rose** [comment]
Closed since it's not ready.

**Pull request closed without merging**